data_diff/diff_tables.py

```diff
@@ -211,14 +223,18 @@ def _diff_tables(self, table1, table2, level=0):
         # Create new instances of TableSegment between each checkpoint
         segmented1 = table1.segment_by_checkpoints(mutual_checkpoints)
         segmented2 = table2.segment_by_checkpoints(mutual_checkpoints)
+
         if self.debug:
             logger.debug("Performing sanity tests for chosen segments (assert sum of fragments == whole)")
+            precalc_attr("count", segmented1 + segmented2)
             assert count1 == sum(s.count for s in segmented1)
             assert count2 == sum(s.count for s in segmented2)
 
-        # Compare each pair of corresponding segments between table1 and table2
-        for i, (t1, t2) in enumerate(safezip(segmented1, segmented2)):
-            logger.info(". " * level + f"Diffing segment {i+1}/{len(segmented1)} of size {t1.count} and {t2.count}")
-            if t1.checksum != t2.checksum:
-                # Apply recursively
-                yield from self._diff_tables(t1, t2, level + 1)
+        # Recursively compare each pair of corresponding segments between table1 and table2
+        diff_iters = [
+            self._diff_tables(t1, t2, level + 1, i + 1, len(segmented1))
+            for i, (t1, t2) in enumerate(safezip(segmented1, segmented2))
+        ]
+
+        for res in thread_map(list, diff_iters):

```

Comment:
I wonder whether it's worth it to wait with printing the verbose output until this point too... so you can sort it by segment. I see pros and cons with each...

Reply:
<img width="1211" alt="CleanShot 2022-05-19 at 10 38 47@2x" src="https://user-images.githubusercontent.com/97400/169322974-64c0081f-d551-40d7-bdec-8ea629e474cb.png">

Reply:
Yeah, I agree... I don't think this is clear... it at least makes it quite clear that it's threaded. I don't think we need to change it